### ISSUE TITLE
Add linter to find unused code

### ISSUE THREAD

**Issue by AlekseySh** [OML-Team/open-metric-learning#109]

**Comment by fk00750** [comment]
Hey can you assign me this issue

**Comment by AlekseySh** [comment]
@fk00750 I am really sorry, I just see this comment. Are you still interested in working on this issue? We still need to complete this issue (and others as well) :)

**Comment by Illumaria** [comment]
Hello! Another wild interested person appeared. 😀 Would you prefer a faster but less precise static linter, or a slower but more reliable dynamic one (like `coverage` that works during tests run)?

**Comment by AlekseySh** [comment]
@Illumaria Hi! Thank you for your interest. Ideally, we want to see suck a linter as a part of pre-commit hooks. But putting it into the tests is also okay, but less preferable.

By the way, when does static one work? Not during the test run?

**Comment by Illumaria** [comment]
> @Illumaria Hi! Thank you for your interest. Ideally, we want to see suck a linter as a part of pre-commit hooks. But putting it into the tests is also okay, but less preferable.
> 
> By the way, when does static one work? Not during the test run?

I don't want to take credit for someone else's merits, so here's the [answer](https://stackoverflow.com/a/9824998/11241660) that covers the differences. Unlike dynamic ones, static code analyzers don't need to run any code, so they can be perfectly added as a pre-commit hook and work outside of a test run.

**Comment by AlekseySh** [comment]
@Illumaria cool, then I would say let's pick some of the static ones and use it as a pre-commit hook

**Comment by Illumaria** [comment]
Alright, so I decided to just go ahead and try them all. So far I've either tried or came across:
- [`vulture`](https://pypi.org/project/vulture/): see below;
- [`pyflakes`](https://pypi.org/project/pyflakes/): see below;
- [`dead`](https://pypi.org/project/dead/): too basic by its nature, thus can't be better than the ones above;
- [`uncalled`](https://pypi.org/project/uncalled/): outdated;
- [`pychecker`](https://pypi.org/project/PyChecker/): very outdated.

I've also checked whether [`Pylint`](https://pypi.org/project/pylint/) has an option of detecting uncalled functions - alas, the devs [rejected](https://github.com/PyCQA/pylint/issues/5773) this feature request, but suggested to use other third-party tools like `vulture`.

---

1. [`vulture`](https://pypi.org/project/vulture/):
    Pros:
    - has an option of specifying `min-confidence` threshold:
    ```commandline
    > python -m vulture --min-confidence=80 .\oml 
    oml\lightning\callbacks\metric.py:59: unused variable 'batch_idx' (100% confidence)
    oml\lightning\callbacks\metric.py:75: unused variable 'batch_idx' (100% confidence)   
    oml\lightning\modules\retrieval.py:76: unused variable 'batch_idx' (100% confidence)  
    oml\lightning\modules\retrieval.py:91: unused variable 'batch_idx' (100% confidence)  
    oml\lightning\modules\retrieval.py:91: unused variable 'dataset_idx' (100% confidence)
    ```
    Cons:
    - outputs very low confidence towards definitely unused functions. I've added one myself to `oml/const.py`:
    ```python
    def uncalled_function(x: int) -> int:
        return 0
    ```
    and `vulture` detected it only when I lowered the threshold to 60%, but this also led to a lot of false positives all over the project:
    ```commandline
    > python -m vulture --min-confidence=60 .\oml
    oml\const.py:26: unused variable 'TMP_PATH' (60% confidence)
    oml\const.py:29: unused variable 'CONFIGS_PATH' (60% confidence)                                
    oml\const.py:30: unused variable 'MODELS_CONFIGS_PATH' (60% confidence)                         
    oml\const.py:97: unused function 'uncalled_function' (60% confidence)                           
    oml\lightning\callbacks\metric.py:58: unused method 'on_validation_batch_start' (60% confidence)
    oml\lightning\callbacks\metric.py:59: unused variable 'batch_idx' (100% confidence)             
    oml\lightning\callbacks\metric.py:69: unused method 'on_validation_batch_end' (60% confidence)  
    oml\lightning\callbacks\metric.py:75: unused variable 'batch_idx' (100% confidence)             
    oml\lightning\callbacks\metric.py:111: unused method 'on_validation_epoch_end' (60% confidence) 
    oml\lightning\callbacks\metric.py:184: unused method 'on_train_start' (60% confidence)          
    oml\lightning\callbacks\metric.py:187: unused method 'on_validation_start' (60% confidence)     
    oml\lightning\modules\retrieval.py:76: unused method 'training_step' (60% confidence)           
    oml\lightning\modules\retrieval.py:76: unused variable 'batch_idx' (100% confidence)            
    oml\lightning\modules\retrieval.py:91: unused method 'validation_step' (60% confidence)
    oml\lightning\modules\retrieval.py:91: unused variable 'batch_idx' (100% confidence)
    oml\lightning\modules\retrieval.py:91: unused variable 'dataset_idx' (100% confidence)
    oml\lightning\modules\retrieval.py:95: unused method 'configure_optimizers' (60% confidence)
    oml\lightning\modules\retrieval.py:114: unused method 'on_epoch_start' (60% confidence)
    oml\models\resnet.py:87: unused attribute 'avgpool' (60% confidence)
    oml\models\resnet.py:141: unused method 'draw_gradcam' (60% confidence)
    oml\models\vit\hubconf.py:22: unused variable 'dependencies' (60% confidence)
    oml\models\vit\hubconf.py:85: unused function 'dino_resnet50' (60% confidence)
    oml\models\vit\hubconf.py:101: unused function 'dino_xcit_small_12_p16' (60% confidence)
    oml\models\vit\hubconf.py:115: unused function 'dino_xcit_small_12_p8' (60% confidence)
    oml\models\vit\hubconf.py:129: unused function 'dino_xcit_medium_24_p16' (60% confidence)
    oml\models\vit\hubconf.py:143: unused function 'dino_xcit_medium_24_p8' (60% confidence)
    oml\models\vit\vision_transformer.py:187: unused variable 'H' (60% confidence)
    oml\models\vit\vision_transformer.py:187: unused variable 'W' (60% confidence)
    oml\models\vit\vision_transformer.py:214: unused attribute 'num_features' (60% confidence)
    oml\models\vit\vision_transformer.py:245: unused attribute 'head' (60% confidence)
    oml\models\vit\vision_transformer.py:283: unused variable 'nc' (60% confidence)
    oml\models\vit\vision_transformer.py:311: unused method 'get_intermediate_layers' (60% confidence)
    oml\models\vit\vision_transformer.py:322: unused function 'vit_tiny' (60% confidence)
    oml\models\vit\vision_transformer.py:336: unused function 'vit_small' (60% confidence)
    oml\models\vit\vision_transformer.py:350: unused function 'vit_base' (60% confidence)
    oml\models\vit\vision_transformer.py:364: unused class 'DINOHead' (60% confidence)
    oml\models\vit\vit.py:138: unused method 'draw_attention' (60% confidence)
    oml\registry\__init__.py:12: unused function 'show_registry' (60% confidence)
    oml\samplers\category_balance.py:72: unused attribute '_resample_labels' (60% confidence)
    oml\transforms\images\albumentations\transforms.py:71: unused class 'Crop' (60% confidence)
    oml\utils\misc.py:37: unused attribute 'benchmark' (60% confidence)
    oml\utils\misc.py:38: unused attribute 'deterministic' (60% confidence)
    ```
    In case of `vulture` this can be fought by generating whitelists/excluding files, but it seems like an overkill.

2. [`pyflakes`](https://pypi.org/project/pyflakes/):
    Pros:
    - works 🤷🏻‍♂️

    Cons:
    - has literally no configuration;
    - emits false positives;
    - _very_ prudent -> ignores definitely unused functions (same experiment with `uncalled_function` as above):
    ```commandline
    > python -m pyflakes .\oml                    
    .\oml\interfaces\datasets.py:7:1: 'oml.const.INPUT_TENSORS_KEY' imported but unused
    .\oml\interfaces\datasets.py:7:1: 'oml.const.IS_GALLERY_KEY' imported but unused            
    .\oml\interfaces\datasets.py:7:1: 'oml.const.IS_QUERY_KEY' imported but unused              
    .\oml\interfaces\datasets.py:7:1: 'oml.const.LABELS_KEY' imported but unused                
    .\oml\interfaces\datasets.py:13:1: 'oml.samplers.balance.BalanceSampler' imported but unused
    ```

---

To sum it up, it is easy to add `vulture` with some high threshold (to fight false positives) to pre-commit hooks, but the overall impact of this is very questionable. I would probably suggest using per-file coverage analysis/blockage.

**Comment by AlekseySh** [comment]
@Illumaria Thank you for such a detailed comparison! I have one question. Is it possible to ignore one particular line in vulture? (like `type: ignore` in mypy)

**Comment by AlekseySh** [comment]
@Illumaria got it
